dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java

```diff
@@ -0,0 +1,167 @@
+        private NullCheckReplacer(SignChecker signChecker) {
+            this.signChecker = signChecker;
+        }
+
+        @Override
+        public Expression visitIntCmpExpression(IntCmpExpr cmp) {
+            if (cmp.getRight() instanceof IntLiteral lit && lit.isZero() && cmp.getLeft().accept(signChecker) == Sign.POS) {
+                // Simplify "expr cop 0" if <expr> is known to be positive.
+                final IntCmpOp op = cmp.getKind();
+                final boolean valueOnNonNull = op == NEQ || op == GT || op == GTE || op == UGT || op == UGTE;
+                return ExpressionFactory.getInstance().makeValue(valueOnNonNull);
+            }
+            return cmp;
+        }
+    }
+
+    private static class SignChecker implements ExpressionVisitor<Sign>, EventVisitor<Void> {
+
+        private final Map<Register, Sign> signMap = new HashMap<>();
+
+        @Override
+        public Void visitEvent(Event e) {
+            if (e instanceof RegWriter writer) {
+                signMap.put(writer.getResultRegister(), Sign.UNKNOWN);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitLocal(Local e) {
+            final Sign sign = e.getExpr().accept(this);
+            signMap.compute(e.getResultRegister(), (key, s) -> s == null ? sign : Sign.meet(s, sign));
+            return null;
+        }
+
+        @Override
+        public Void visitAlloc(Alloc e) {
+            final Sign sign = Sign.POS;
+            signMap.compute(e.getResultRegister(), (key, s) -> s == null ? sign : Sign.meet(s, sign));
+            return null;
+        }
+
+
+        // ============================== Expressions ==============================
+
+        @Override
+        public Sign visitExpression(Expression expr) {
+            return Sign.UNKNOWN;
+        }
+
+        @Override
+        public Sign visitRegister(Register reg) {
+            return signMap.getOrDefault(reg, Sign.UNKNOWN);
+        }
+
+        @Override
+        public Sign visitMemoryObject(MemoryObject memObj) {
+            return Sign.POS;
+        }
+
+        @Override
+        public Sign visitFunction(Function function) {
+            return Sign.POS;
+        }
+
+        @Override
+        public Sign visitIntLiteral(IntLiteral lit) {
+            final int cmpRes = lit.getValue().compareTo(BigInteger.ZERO);
+            return cmpRes > 0 ? Sign.POS : cmpRes == 0 ? Sign.NON_NEG : Sign.UNKNOWN;
+        }
+
+        @Override
+        public Sign visitIntBinaryExpression(IntBinaryExpr expr) {
+            final Sign leftSign = expr.getLeft().accept(this);
+            final Sign rightSign = expr.getRight().accept(this);
+            if (leftSign == Sign.UNKNOWN || rightSign == Sign.UNKNOWN) {
+                return Sign.UNKNOWN;
+            }
+
+            // --- Both subexpressions are (at least) non-negative ---
+            final IntBinaryOp op = expr.getKind();
+            if (leftSign == Sign.POS && op == ADD) {
```

Comment:
Should this be `(leftSign == Sign.POS || rightSign == Sign.POS) && op == ADD`?

Reply:
I guess that would be slightly more accurate